File: abc/abc166/abc166_e.fs
// https://atcoder.jp/contests/abc166/submissions/12815277

let N = stdin.ReadLine() |> int
let A = stdin.ReadLine().Split() |> Array.map int64

let d = Array.zeroCreate<int64> N
let s = Array.zeroCreate<int64> N

let calc ope cond (a : int64 []) =
    A
    |> Array.indexed
    |> Array.filter (fun x -> cond ((ope) (fst x) (snd x |> int)))
    |> Array.map (fun x -> (ope) (fst x) (snd x |> int))
    |> Array.groupBy id
    |> Array.map (fun x -> (fst x, Array.length (snd x)))
    |> Array.map (fun x -> a.[fst x] <- snd x |> int64)
    |> ignore

calc (-) (fun x -> x >= 0) d
calc (+) (fun x -> x < N) s

Array.zip d s
|> Array.map (fun x -> fst x * snd x)
|> Array.sum
|> stdout.WriteLine

File: abc/abc166/abc166_e.cs
// https://atcoder.jp/contests/abc166/submissions/12812585

using System;
using System.Linq;

class Program
{
    static void Main()
    {
        var N = long.Parse(Console.ReadLine());
        var A = Console.ReadLine().Split().Select(long.Parse).ToArray();

        var sum = new long[N];
        var diff = new long[N];

        for (var i = 0; i < N; i++)
        {
            if (i - A[i] >= 0)
            {
                diff[i - A[i]]++;
            }
            if (i + A[i] < N)
            {
                sum[i + A[i]]++;
            }
        }
        Console.WriteLine(sum.Zip(diff, (s, d) => s * d).Sum());
    }
}
